File: bada_ruby/lib/bada/qc/logic.rb
# frozen_string_literal: true

module Bada
  module QC
    # Logic — the *semiconductor* electronic-circuit simulation that drives the
    # classical control of the pseudo quantum computer.
    #
    # 「電子回路をシミュレーションで作って…半導体の…」——CMOS の MOSFET しきい値
    #   モデルから NAND を作り、NAND だけからインバータ／AND／OR／XOR と n-bit
    #   デコーダを組み上げます。CPU の命令デコードは、この*シミュレーション結果*で
    #   ワンホット選択を行います（＝半導体回路の結果を実際に使う）。
    #
    # Logic levels are volts on [0, VDD]; a MOSFET conducts above threshold VTH.
    module Logic
      module_function

      VDD = 1.0
      VTH = 0.5

      def high?(v)
        v >= VTH
      end

      def level(bit)
        bit == 1 || bit == true ? VDD : 0.0
      end

      # CMOS NAND: output low only when both inputs pull the series nFETs on.
      def nand(a, b)
        (high?(a) && high?(b)) ? 0.0 : VDD
      end

      # Inverter (NOT) = NAND with tied inputs.
      def inv(a)
        nand(a, a)
      end

      def and_(a, b)
        inv(nand(a, b))
      end

      def or_(a, b)
        nand(inv(a), inv(b))
      end

      def xor_(a, b)
        # xor = (a NAND (a NAND b)) NAND (b NAND (a NAND b))
        n = nand(a, b)
        nand(nand(a, n), nand(b, n))
      end

      # Read the k-th bit of value as a logic level.
      def bit_level(value, k)
        level((value >> k) & 1)
      end

      # An n-bit one-hot decoder built purely from NAND/INV gates. Returns an
      # array of 2**bits logic levels; exactly one is high (VDD). This is run by
      # the CPU to select which gate microcode to execute.
      def decode(value, bits)
        lines = (0...bits).map { |k| bit_level(value, k) }
        (0...(1 << bits)).map do |sel|
          # AND together each address line (or its complement) — an n-input AND
          # implemented as a NAND tree followed by an inverter.
          acc = VDD
          (0...bits).each do |k|
            want = (sel >> k) & 1
            line = want == 1 ? lines[k] : inv(lines[k])
            acc = and_(acc, line)
          end
          acc
        end
      end

      # Index of the (single) high output of the decoder — the decoded opcode.
      def decode_index(value, bits)
        decode(value, bits).index { |v| high?(v) } || 0
      end
    end
  end
end

File: apps/core/src/main/java/bada/qc/Logic.java
package bada.qc;

/**
 * Logic — the semiconductor electronic-circuit simulation driving the classical
 * control: CMOS/MOSFET NAND, the gates derived from it, and a one-hot opcode
 * decoder used by the CPU. Java port of Bada::QC::Logic.
 */
public final class Logic {
    public static final double VDD = 1.0;
    public static final double VTH = 0.5;

    public static boolean high(double v) { return v >= VTH; }
    public static double level(int bit) { return bit == 1 ? VDD : 0.0; }

    public static double nand(double a, double b) {
        return (high(a) && high(b)) ? 0.0 : VDD;
    }

    public static double inv(double a) { return nand(a, a); }
    public static double and(double a, double b) { return inv(nand(a, b)); }
    public static double or(double a, double b) { return nand(inv(a), inv(b)); }

    public static double xor(double a, double b) {
        double n = nand(a, b);
        return nand(nand(a, n), nand(b, n));
    }

    public static double bitLevel(int value, int k) {
        return level((value >> k) & 1);
    }

    /** One-hot decoder from NAND/INV; returns 2^bits levels, exactly one high. */
    public static double[] decode(int value, int bits) {
        double[] lines = new double[bits];
        for (int k = 0; k < bits; k++) lines[k] = bitLevel(value, k);
        double[] out = new double[1 << bits];
        for (int sel = 0; sel < (1 << bits); sel++) {
            double acc = VDD;
            for (int k = 0; k < bits; k++) {
                int want = (sel >> k) & 1;
                double line = want == 1 ? lines[k] : inv(lines[k]);
                acc = and(acc, line);
            }
            out[sel] = acc;
        }
        return out;
    }

    public static int decodeIndex(int value, int bits) {
        double[] d = decode(value, bits);
        for (int i = 0; i < d.length; i++) if (high(d[i])) return i;
        return 0;
    }

    private Logic() { }
}
